aspen/http/resource.py

```diff
@@ -29,25 +32,66 @@ class Dynamic(Simplate):
     """Model a dynamic HTTP resource using simplates.
 
        Make .request_processor available as it has been historically.
-
-       Figure out which accept header to use.
-
     """
 
-    def __init__(self, request_processor, fs, raw, default_media_type):
+    def __init__(self, request_processor, fs, raw, fs_media_type):
         self.request_processor = request_processor
+        self.fs_media_type = fs_media_type
         defaults = request_processor.simplate_defaults
+        default_media_type = fs_media_type or request_processor.media_type_default
         super(Dynamic, self).__init__(defaults, fs, raw, default_media_type)
 
-
     def render(self, state):
+        """Render the resource with the given state as context, return Output.
+
+        Before rendering we need to determine what type of content we're going
+        to send back, by trying to find a match between the media types the
+        client wants and the ones provided by the resource.
+
+        The two sources for what the client wants are the extension in the
+        request URL, and the Accept header. If the former fails to match we
+        raise NotFound (404), if the latter fails we raise NegotiationFailure
+        (406).
+
+        Note that we don't always respect the `Accept` header (the spec says
+        we can ignore it: <https://tools.ietf.org/html/rfc7231#section-5.3.2>).
+        """
+        available = self.available_types
+        choice = 0  # default to first available type
+        # When there is an extension in the URI path, the dispatcher gives us the
+        # corresponding media type (or an empty string if unknown)
         accept = dispatch_accept = state['dispatch_result'].extra.get('accept')
-        if accept is None:
-            accept = state.get('accept_header')
-        try:
-            return super(Dynamic, self).render(accept, state)
-        except NegotiationFailure:
-            if dispatch_accept is not None:
-                # e.g. client requested `/foo.json` but `/foo.spt` has no JSON page
+
+        if dispatch_accept is not None:
+            # If the extension is unknown, raise NotFound
```

Comment:
Hmm, I think the previous behavior was inconsistent, as `dispatch_accept` would be `None`, so `accept` would be set to `accept_header`, and thus it could: return a 406, or choose a type from the Accept header which that doesn't match the extension in the URL path, or fall back to the first available type.

I think we do want to return a 404. If the developer wants to use a custom extension they can register it in the `mimetypes` module.